### PR TITLE
Depth-recurrent transformer: shared block × 12 passes, val_bpb 1.4061, 4.39MB artifact

### PR DESCRIPTION
**Pull request by Sambhav242005**
## Approach
Single shared transformer block applied 12 times instead of independent layers.
Reduces unique params from ~25M → 4.7M, freeing budget to go wider (dim=768).

Key changes from baseline:
- Depth recurrence: 1 shared Block × num_passes (default 12)
- U-Net skip connections: encoder passes store activations, decoder passes consume in reverse
- x0 residual mix: original embedding injected at every pass for stability
- Wider model: MODEL_DIM=768, NUM_HEADS=12 vs baseline 512/8

## Result
- val_bpb: 1.4061 (after int8 + zlib roundtrip)
- Artifact size: 4.39MB / 16MB budget
- Parameters: 4,726,284
- Hardware: RTX 4090, ~45s training time

## Run command
MODEL_DIM=768 NUM_HEADS=12 NUM_KV_HEADS=4 NUM_PASSES=12 
MATRIX_LR=0.02 WARMDOWN_ITERS=600 MAX_WALLCLOCK_SECONDS=0 
TRAIN_BATCH_TOKENS=524288 PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
torchrun --standalone --nproc_per_node=1 train_gpt_recurrent.py

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Depth-recurrent transformer: shared block × 12 passes, val_bpb 1.4061, 4.39MB artifact

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


Head SHA: 91fc1bf897e8bb4c5140b160dd133b31d3057ce6

## Check 1: N-gram family bug (CLOSE trigger)

No n-gram hash structures present. Searched the full file for `ngram`, `n_gram`, `BigramHash`, `TrigramHash`, `NgramHash`, `bigram`, and `hash`. Zero matches. The scoring pipeline uses only a sentencepiece LUT (`build_sentencepiece_luts`) for byte accounting — this is a static lookup by token ID, not a hash keyed on (prev_token, target_token) pairs. No CLOSE trigger.

## Check 2: Pre-Quant TTT bug (CLOSE trigger)

`_eval_ttt` is present and enabled by default (`TTT_ENABLED=1`). The function uses `torch.optim.Adam` (not AdamW) on all model parameters. Critically, examining the loop:

```python
# Step 1: score with no_grad → val_loss_sum accumulated
with torch.no_grad(), torch.autocast(...):
    score_loss = base_model(x, y).detach()
val_loss_sum += score_loss.to(torch.float64) * batch_token_count
val_token_count += batch_token_count
val_byte_count += _count_bytes(...)

# Step 2: train on same chunk
for _ttt_step in range(args.ttt_steps):
    ttt_optimizer.zero_grad()
    with torch.autocast(...):
        ttt_loss = base_model(x, y)
    ttt_loss.backward()
    ttt_optimizer.step()
```

Score is recorded first under `no_grad` before any gradient update on that chunk. This is the canonical score-first-per-chunk pattern. The bug pattern (CLOSE) requires multi-epoch AdamW on val_tokens without score-first — neither condition applies here. Optimizer is Adam (not AdamW), and score-first ordering is correctly implemented. No CLOSE trigger.

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by MatoTeziTanka** [comment]
# PR 386: Depth-Recurrent Transformer

## Review Summary

**PR Title:** Depth-recurrent transformer: shared block × 12 passes, val_bpb 1.4061, 4.39MB artifact
**Status:** OPEN | No reviewer comments
**Train File:** train_gpt.py (root level, main branch modification)
**Classification:** PURE_NEURAL_CLEAN

## Red Flag Analysis

| Signal | Finding |
|--------|---------|
| target-in-key loss | CLEAN - Uses tokenizer-agnostic BPB metric |
| TTT/SLOT classes | CLEAN - No TTT/SLOT classes found |
| Custom tokenizer | CLEAN - Only standard SentencePiece tokenizer |
| loss[] indexing | CLEAN - No custom loss dict access patterns |

## Findings

1. **Main branch modification**: Changes baseline train_gpt.py to support depth-recurrent stacking
2. **Approach**: Single transformer block recycled through 12 sequential passes (parameter sharing)
3. **Result**: val_bpb=1.4061 (higher BPB than records, but artifact is only 4.39MB - extreme compression exploration)
4. **Note**: This is baseline-modification PR, standardizes depth-recurrent architecture for future submissions

## Recommendation

**MERGE** - Legitimate architectural exploration. Parameter sharing (depth recurrence) is well-established technique. No loss gaming or tokenizer manipulation. Useful baseline enhancement.


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
